src/worker/rayon.rs

```diff
@@ -0,0 +1,182 @@
+        let quantization_table;
+        let block_count;
+        let line_stride;
+        let block_size;
+        let dct_scale;
+
+        {
+            let inner = mutex.lock().unwrap();
+            quantization_table = inner.quantization_tables[index].as_ref().unwrap().clone();
+            block_size = inner.components[index].as_ref().unwrap().block_size;
+            let metadata = inner.component_metadata(index);
+
+            block_count = metadata.block_count;
+            line_stride = metadata.line_stride;
+            dct_scale = metadata.dct_scale;
+        }
+
+        assert_eq!(data.len(), block_count * 64);
+
+        let mut output_buffer = [0; 64];
+        for i in 0..block_count {
+            let x = (i % block_size.width as usize) * dct_scale;
+            let y = (i / block_size.width as usize) * dct_scale;
+
+            let coefficients: &[i16; 64] = &data[i * 64..(i + 1) * 64].try_into().unwrap();
+
+            // Write to a temporary intermediate buffer, a 8x8 'image'.
+            dequantize_and_idct_block(
+                dct_scale,
+                coefficients,
+                &*quantization_table,
+                8,
+                &mut output_buffer,
+            );
+
+            // Lock the mutex only for this write back, not the main computation.
+            // FIXME: we are only copying image data. Can we use some atomic backing buffer and a
+            // `Relaxed` write instead?
+            let mut write_back = mutex.lock().unwrap();
+            let write_back = &mut write_back.results[index][result_offset + y * line_stride + x..];
+
+            let buffered_lines = output_buffer.chunks_mut(8);
+            let back_lines = write_back.chunks_mut(line_stride);
+
+            for (buf, back) in buffered_lines.zip(back_lines).take(dct_scale) {
+                back[..dct_scale].copy_from_slice(&buf[..dct_scale]);
+            }
+        }
+    }
+}
+
+impl super::Worker for Scoped {
+    fn start(&mut self, row_data: RowData) -> Result<()> {
+        self.inner.get_mut().unwrap().start_immediate(row_data);
+        Ok(())
+    }
+
+    fn append_row(&mut self, row: (usize, Vec<i16>)) -> Result<()> {
+        let (index, data) = row;
+        let result_offset;
+
+        {
+            let mut inner = self.inner.get_mut().unwrap();
+            let metadata = inner.component_metadata(index);
+
+            result_offset = inner.offsets[index];
+            inner.offsets[index] += metadata.bytes_used();
+        }
+
+        ImmediateWorker::append_row_locked(&self.inner, (index, data), result_offset);
+        Ok(())
+    }
+
+    fn get_result(&mut self, index: usize) -> Result<Vec<u8>> {
+        let result = self.inner.get_mut().unwrap().get_result_immediate(index);
+        Ok(result)
+    }
+
+    // Magic sauce, these _may_ run in parallel.
+    fn append_rows(&mut self, iter: &mut dyn Iterator<Item = (usize, Vec<i16>)>) -> Result<()> {
+        rayon::in_place_scope(|scope| {
```

Comment:
I think you can drop the mutex everywhere except for the final `write_back` above if you iterate over the set of tasks twice. The only writes in `append_row_locked` are to the `results` array so the locking here and for computing the DCT parameters can be skipped.

Removing the mutex entirely without dropping to unsafe seems possible since the writes don't overlap, but possibly trickier than it's worth. You'd have to benchmark whether keeping track of everything and using slice::split_at_mut would actually be faster than just keeping the lock. If the tasks are sorted, or can be sorted, it'd probably make it easier.

Reply:
Rust lifetimes and borrow checker are amazing when you _really_ use them. We can get rid of mutex entirely, with no unsafe as you said. Tasks are sorted within each component.